integration/utils/worker.go

```diff
@@ -1,42 +1,66 @@
 package utils
 
 import (
+	"context"
+	"sync"
 	"time"
 )
 
+type workFunc func(workerID int)
+
 type Worker struct {
 	workerID int
 	interval time.Duration
-	work     func(workerID int)
-	ticker   *time.Ticker
-	quit     chan struct{}
+	work     workFunc
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	wg *sync.WaitGroup
 }
 
-func NewWorker(workerID int, interval time.Duration, work func(workerID int)) Worker {
+func NewWorker(workerID int, interval time.Duration, work workFunc) Worker {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
 	return Worker{
 		workerID: workerID,
 		interval: interval,
 		work:     work,
+
+		ctx:    ctx,
+		cancel: cancel,
+
+		wg: wg,
 	}
 }
 
 func (w *Worker) Start() {
-	w.ticker = time.NewTicker(w.interval)
-	w.quit = make(chan struct{})
-
 	go func() {
-		for {
+		defer w.wg.Done()
```

Comment:
why do we need this one? we don't even have a corresponding `w.wg.Add(1)` for this

Reply:
it's a tricky one.  Initially, this wg was split into two: one for all the workers (to prove that they exited) and one for the loop (to prove that no new workers would be submitted).   Then I've simplified that by coalescing them into one, and doing `wg.Add(1)` at the creation of the worker.   Let me move it to the Start() function instead so that it'll be move obvious.